presto-connector-tests/src/main/java/com/facebook/presto/connector/unittest/BaseMetadataTest.java

```diff
@@ -43,43 +44,121 @@
     }
 
-    default List<String> systemSchemas()
+    /*
+     * Connectors that add columns containing connector metadata to a table, like Hive,
+     * should return a List containing the metadata for all of the expectedColumns and the connector-specific columns added in the appropriate positions.
+     */
+    default List<ColumnMetadata> extendWithConnectorSpecificColumns(List<ColumnMetadata> expectedColumns)
     {
-        return ImmutableList.of();
+        return expectedColumns;
     }
 
+    /*
+     * Returns a SchemaTableName to be used when the name of the schema is NOT
+     * important.
+     *
+     * Tests that require tables in one schema should use this method. An
+     * example would be a test that renames a table within a single schema.
+     *
+     * Tests written using this can be run against connectors that support the
+     * creation of schemas and those that do not, but come with a usable schema
+     * out of the box. PostgreSQL, for example has a "public" schema.
+     *
+     * A class implementing this interface for e.g. PostgreSQL should return a
+     * SchemaTableName with the schemaName set to "public".
+     *
+     * Tests invoking this method do NOT generally need to be annotated with
+     * @RequiredFeatures({CREATE_SCHEMA, DROP_SCHEMA}).
+     *
+     * A class overriding this method should also override withSchemas, as
+     * described below.
+     */
     default SchemaTableName schemaTableName(String tableName)
     {
         return new SchemaTableName("default_schema", tableName);
     }
 
+    /*
+     * Returns a SchemaTableName to be used when the name of the schema is
+     * important.
+     *
+     * Tests that require tables in multiple schemas should use this method. An
+     * example would be a test that renames a table from one schema to another.
+     *
+     * Tests invoking this method must be annotated with
+     * @RequiredFeatures({CREATE_SCHEMA, DROP_SCHEMA})
+     */
     default SchemaTableName schemaTableName(String schemaName, String tableName)
     {
         return new SchemaTableName(schemaName, tableName);
     }
 
+    /*
+     * Create and drop the schemas specified by schemaNames, running callable
+     * in between creating and dropping the schemas.
+     *
+     * A class implementing this interface for a connector that does NOT
+     * support creating and/or dropping schemas, but DOES have a usable schema
+     * (e.g. PostgreSQL with the "public" schema) should override this method
+     * and invoke callable without trying to create or destroy any of the
+     * schemas specified.
+     *
+     * Overriding this method in such a way allows tests that explicitly
+     * require creating and dropping tables to be written such that they are
+     * independent of whether or not the connector supports creating and
+     * dropping schemas.
+     */
+    default void withSchemas(ConnectorSession session, List<String> schemaNames, Callable<Void> callable)
```

Comment:
Is an additional `return null` that bad compared to introducing an extra interface?
I'd probably prefer using standard interfaces as long as they fit (and an additional `return null` does not qualify for `not fit` to me).

Reply:
I disagree that we should have withMetadata() methods with some number of Consumers<ConnectorMetadata>. It's a lot of boilerplate code to remove 3 instances of ImmutableList.of(). Which are lists of 2, 5, and 6 calls; should we implement the ones for 1, 3, and 4 calls in case we need them in the future?

I further disagree that we should do something like this with withSchemas and withTables. Most of the existing code that uses them is written in terms of the methods that take List<Consumer<Metadata>>. Should we have the cross product of number of schemas/tables and number of consumers just to save a couple ImmutableList.of() calls?